### PR TITLE
Add start date to trigger_dagrun operator

### LINKED ISSUE #18082: TriggerDagRunOperator start_date is not set
### Apache Airflow version

2.1.3 (latest released)

### Operating System

ubuntu, macos

### Versions of Apache Airflow Providers

apache-airflow-providers-amazon==1.4.0
apache-airflow-providers-ftp==1.1.0
apache-airflow-providers-http==1.1.1
apache-airflow-providers-imap==1.0.1
apache-airflow-providers-postgres==1.0.2
apache-airflow-providers-slack==4.0.0
apache-airflow-providers-sqlite==1.0.2

### Deployment

Virtualenv installation

### Deployment details

_No response_

### What happened

We are using TriggerDagRunOperator in the end of DAG to retrigger current DAG:

`TriggerDagRunOperator(task_id=‘trigger_task’, trigger_dag_id=‘current_dag’)`

Everything works fine, except we have missing duration in UI and ​warnings in scheduler :

`[2021-09-07 15:33:12,890] {dagrun.py:604} WARNING - Failed to record duration of <DagRun current_dag @ 2021-09-07 12:32:17.035471+00:00: manual__2021-09-07T12:32:16.956461+00:00, externally triggered: True>: start_date is not set.`

And in web UI we can't see duration in TreeView and DAG has no started and duration values. 

### What you expected to happen

Correct behaviour with start date and duration metrics in web UI.

### How to reproduce

```
import pendulum
import pytz
from airflow import DAG
from airflow.operators.dummy import DummyOperator
from airflow.operators.trigger_dagrun import TriggerDagRunOperator


with DAG(
        start_date=pendulum.datetime(year=2021, month=7, day=1).astimezone(pytz.utc),
        schedule_interval='@daily',
        default_args={},
        max_active_runs=1,
        dag_id='current_dag'
) as dag:

    step1 = DummyOperator(task_id='dummy_task')

    trigger_self = TriggerDagRunOperator(task_id='trigger_self', trigger_dag_id='current_dag')

    step1 >> trigger_self
```

`[2021-09-08 12:53:35,094] {dagrun.py:604} WARNING - Failed to record duration of <DagRun current_dag @ 2021-01-04 03:00:11+00:00: backfill__2021-01-04T03:00:11+00:00, externally triggered: False>: start_date is not set.`

### Anything else

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by bhavaniravi**
closes: #18082  

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

**Review by ephraimbuddy — CHANGES_REQUESTED** [review]
I think that you made the change in the wrong place. Here's where I think the change should be:
https://github.com/apache/airflow/blob/23a68fa4e61ba795340f095ab64fbcaeebd57db4/airflow/operators/trigger_dagrun.py#L132
Add `start_date=self.start_date` after the line above.

EDIT:
You are actually correct, just fix the failing tests

**Comment by ephraimbuddy** [comment]
> I think that you made the change in the wrong place. Here's where I think the change should be:
> https://github.com/apache/airflow/blob/23a68fa4e61ba795340f095ab64fbcaeebd57db4/airflow/operators/trigger_dagrun.py#L132
> 
> 
> Add `start_date=self.start_date` after the line above.

But it seems the trigger_dag doesn't take start_date argument

**Review by ephraimbuddy — APPROVED** [review]

File: tests/operators/test_trigger_dagrun.py [hunk 1]
```diff
@@ -255,3 +255,34 @@ def test_trigger_dagrun_with_wait_for_completion_true_fail(self):
...
+            task_id="test_task",
+            trigger_dag_id=self.dag.dag_id,
+            execution_date=execution_date,
+            allowed_states=[State.RUNNING, State.SUCCESS],
+            dag=self.dag,
+        )
+        with pytest.raises(DagRunAlreadyExists):
+            task.run(start_date=execution_date, end_date=execution_date)
+
+
```
Comment (kaxil) [inline tests/operators/test_trigger_dagrun.py]:
```suggestion
```

File: tests/operators/test_trigger_dagrun.py [hunk 2]
```diff
@@ -255,3 +255,33 @@ def test_trigger_dagrun_with_wait_for_completion_true_fail(self):
         )
         with pytest.raises(AirflowException):
             task.run(start_date=execution_date, end_date=execution_date)
+
```
Comment (kaxil) [inline tests/operators/test_trigger_dagrun.py:258]:
```suggestion
```

File: tests/operators/test_trigger_dagrun.py [hunk 3]
```diff
@@ -255,3 +255,33 @@ def test_trigger_dagrun_with_wait_for_completion_true_fail(self):
...
+        task = TriggerDagRunOperator(
+            task_id="test_task",
+            trigger_dag_id=self.dag.dag_id,
+            execution_date=execution_date,
+            allowed_states=[State.RUNNING, State.SUCCESS],
+            dag=self.dag,
+        )
+        with pytest.raises(DagRunAlreadyExists):
+            task.run(start_date=execution_date, end_date=execution_date)
+
```
Comment (kaxil) [inline tests/operators/test_trigger_dagrun.py]:
```suggestion
```

**Comment by kaxil** [comment]
```
=========================== short test summary info ============================
  FAILED tests/models/test_cleartasks.py::TestClearTasks::test_clear_task_instances
  FAILED tests/models/test_cleartasks.py::TestClearTasks::test_clear_task_instances_with_task_reschedule
  FAILED tests/models/test_cleartasks.py::TestClearTasks::test_dag_clear - asse...
  = 3 failed, 2445 passed, 61 skipped, 3 xfailed, 4 warnings in 508.01s (0:08:28) =
```

tests are failing

**Comment by bhavaniravi** [comment]
Hey, @kaxil, I cannot recreate the failing tests using breeze env or pytest with the same python and Postgres versions. Is there an alternative way to simulate the CI builds?

**Comment by ephraimbuddy** [comment]
Looks like something went wrong in your last rebase. Please rebase again

File: airflow/api/common/experimental/trigger_dag.py [hunk 4]
```diff
@@ -82,6 +82,7 @@ def _trigger_dag(
     for _dag in dags_to_trigger:
         trigger = _dag.create_dagrun(
             run_id=run_id,
+            start_date=timezone.utcnow(),
```
Comment (alexInhert) [inline airflow/api/common/experimental/trigger_dag.py]: If the dag you trigger is timezone aware does it matter?
I mean you consider here the default timezone but the dag you are triggering may be in a diffrent timezone than the default

Comment (bhavaniravi) [reply airflow/api/common/experimental/trigger_dag.py]: The `start_date` here is not the DAG start_date. It is the dag run starting time of the run. Internally it's all tracked in UTC and rendered in the user's timezone on UI.

Correct me if I am wrong.

Comment (uranusjr) [reply airflow/api/common/experimental/trigger_dag.py]: You’re correct. Although maybe we shouldn’t set this to `RUNNING` in the first place, but should set it to `QUEUED` and let the scheduler start the run (and set `start_date`) instead. See how this is implemented in the new REST API (in `airflow/api_connexion/endpoints/dag_run_endpoint.py::post_dag_run`).

Comment (ephraimbuddy) [reply airflow/api/common/experimental/trigger_dag.py]: Supported. Setting it to queued will save us some complaints especially about max_active_runs

Comment (bhavaniravi) [reply airflow/api/common/experimental/trigger_dag.py]: Should I go ahead and make that change? There are 3 places where we are creating dag_runs with the state running, should this be the same behavior across the board?

image

Comment (ephraimbuddy) [reply airflow/api/common/experimental/trigger_dag.py]: We should only change the one identified here. Looks like others are more involved change and requires a separate PR(that's if it should be changed)

Comment (bhavaniravi) [reply airflow/api/common/experimental/trigger_dag.py]: This change will affect three places. All these use the `trigger_dag` function. 

1. TriggerDAGRunOperator
2. APIClient - `dag_runs`
3. CLI trigger dag

File: tests/operators/test_trigger_dagrun.py [hunk 5]
```diff
@@ -255,3 +255,34 @@ def test_trigger_dagrun_with_wait_for_completion_true_fail(self):
...
+            allowed_states=[State.RUNNING, State.SUCCESS],
+            dag=self.dag,
+        )
+        task.run(start_date=execution_date, end_date=execution_date)
+
+        with create_session() as session:
+            dagruns = session.query(DagRun).filter(DagRun.dag_id == self.dag.dag_id).all()
+            assert len(dagruns) == 2
+            dagruns[1].set_state()
+            # assert isinstance(dagruns[1].start_date, datetime)
```
Comment (kaxil) [inline tests/operators/test_trigger_dagrun.py]: Does this need updating?

Comment (kaxil) [reply tests/operators/test_trigger_dagrun.py]: You can just assert that state of 2nd DagRun -- that it is marked as `State.QUEUED`.

**Review by kaxil — COMMENTED** [review]
Unit tests are failing - https://github.com/apache/airflow/pull/18226/checks?check_run_id=3651105087#step:6:6287

**Comment by uranusjr** [comment]
Failure seems unrelated

```
   _____________________ test_external_task_marker_exception ______________________
  
  dag_bag_ext = <airflow.models.dagbag.DagBag object at 0x7fec5c30c730>
  
      def test_external_task_marker_exception(dag_bag_ext):
          """
          Clearing across multiple DAGs should raise AirflowException if more levels are being cleared
          than allowed by the recursion_depth of the first ExternalTaskMarker being cleared.
          """
  >       run_tasks(dag_bag_ext)
  ...
  >       assert task_instance.state == state
  E       AssertionError: assert None == <TaskInstanceState.SUCCESS: 'success'>
  E        +  where None = <TaskInstance: dag_0.task_b_0 manual__2015-01-01T00:00:00+00:00 [None]>.state
```

File: tests/operators/test_trigger_dagrun.py [hunk 6]
```diff
@@ -250,8 +250,36 @@ def test_trigger_dagrun_with_wait_for_completion_true_fail(self):
...
+        execution_date = DEFAULT_DATE
+        task = TriggerDagRunOperator(
+            task_id="test_task",
+            trigger_dag_id=self.dag.dag_id,
+            dag=self.dag,
+        )
+        task.run(start_date=execution_date, end_date=execution_date)
+
+        with create_session() as session:
+            dagruns = session.query(DagRun).filter(DagRun.dag_id == self.dag.dag_id).all()
```
Comment (uranusjr) [inline tests/operators/test_trigger_dagrun.py]: You probably want to add an explicit `order_by` here, the db does not always return rows in a deterministic order.

**Comment by kaxil** [comment]
Thanks @bhavaniravi 🎉

**Pull request merged**